File: backend/api/scheduler.py
# -*- coding:utf-8 -*-
"""
   Copyright 2021 Jerome DE LUCCHI

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

import json
import elasticsearch
import threading
from api import statistic
from api import schedulerManager


class Scheduler:

    def __init__(self, connector):
        self.ES = connector
        self.STATISTIC = statistic.Statistic(self.ES)
        self.STATISTIC_DATA = self.STATISTIC.STATISTIC_DATA
        self.STATISTIC_DATA["object_type"] = "scheduler"
        self.DB_INDEX = 'blast_obj_scheduler'

    def add(self, account_email: str, realm: str, data: dict):
        """ create a new scheduler """

        try:
            data["realm"] = realm
            scheduler_add_res = self.ES.index(index=self.DB_INDEX, body=json.dumps(data), refresh=True)
            if scheduler_add_res["result"] == "created":
                self.STATISTIC_DATA["object_action"] = 'create'
                self.STATISTIC_DATA["object_name"] = data["name"]
                self.STATISTIC_DATA["timestamp"] = statistic.UTC_time()
                self.STATISTIC_DATA["account_email"] = account_email
                self.STATISTIC_DATA["realm"] = realm
                self.STATISTIC.add(self.STATISTIC_DATA)

            return scheduler_add_res

        except Exception as e:
            print("backend Error: file:scheduler:class:scheduler:function:add")
            print(e)
            return {"failure": str(e)}

    def action(self, scheduler_data: dict):

        """ update schedule action """
        try:
            resp_schedule_update = []
            schedules = self.list_by_ids(scheduler_data["realm"], scheduler_data["schedule"])["hits"]["hits"]
            for schedule in schedules:
                schedule["_source"]["status"] = scheduler_data["action"]
                resp_schedule_update.append(self.update(schedule["_id"], schedule["_source"]))
            return resp_schedule_update

        except Exception as e:
            print("backend Error: file:scheduler:class:scheduler:function:action")
            print(e)
            return {"failure": str(e)}

    def execute(self, scheduler_data: dict):

        """
        This function starts a new scheduler thread and execute the scheduler definition
        :param scheduler_data: The scheduler object containing, scheduler_id, realm, account_email
        """
        try:
            schmngr = schedulerManager.SchedulerManager(self.ES)
            schedules = self.list_by_ids(scheduler_data["realm"], scheduler_data["ids"])
            for schedule in schedules["hits"]["hits"]:
                schedule["_source"]["account_email"] = scheduler_data["account_email"]
                new_t = threading.Thread(target=schmngr.execute_schedule, args=(), kwargs=({
                    "scheduler_id": schedule["_id"],
                    "scheduler_source": schedule["_source"]
                }))
                new_t.start()
            return {'success': 'starter'}

        except Exception as e:
            print("backend Exception, file:scheduler:class:scheduler:func:execute")
            print(e)
            return {'failure': str(e)}

    def list(self, realm: str):

        """ list all the scheduler which belong to the provided realm """
        try:
            req = json.dumps({
                "size": 10000,
                "query": {
                    "term": {
                        "realm": realm
                    }
                },
                "sort": [
                    {
                        "name": {
                            "order": "asc"
                        }
                    }
                ]
            })
            return self.ES.search(index=self.DB_INDEX, body=req)

        except (elasticsearch.exceptions.ConnectionError, elasticsearch.exceptions.NotFoundError) as e:
            return {"failure": e}

    def list_all(self):

        """ list all the scheduler no matter the realm it belongs too """
        try:
            req = json.dumps({
                "size": 10000,
                "query": {
                    "match_all": {}
                },
                "sort": [
                    {
                        "name": {
                            "order": "asc"
                        }
                    }
                ]
            })
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print(e)
            return {"failure": str(e)}

    def list_by_scenario(self, realm: str, scenario_id: str):

        try:
            req = json.dumps(
                {
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "realm": {
                                        "term": realm
                                    }
                                },
                                {
                                    "scenario_ids": {
                                        "term": scenario_id
                                    }
                                }
                            ]
                        }
                    }
                }
            )

            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print(e)
            return {"failure": str(e)}


    def list_by_ids(self, realm: str, scheduler_ids: list):

        """ list the schedule for the specific id"""
        try:
            req = json.dumps(
                {
                    "query": {
                        "bool": {
                            "filter": [
                                {
                                    "term": {
                                        "realm": realm
                                    }
                                },
                                {
                                    "terms": {
                                        "_id": scheduler_ids
                                    }
                                }
                            ]
                        }
                    },
                    "sort": [
                        {
                            "name": {
                                "order": "asc"
                            }
                        }
                    ]
                }
            )
            print(req)
            return self.ES.search(index=self.DB_INDEX, body=req)

        except Exception as e:
            print(e)
            return {"failure": e}

    def update(self, schedule_id: str, schedule_data: dict):

        try:
            # self.STATISTIC.__add__(statistic_data)
            return self.ES.update(index=self.DB_INDEX, id=schedule_id, body=json.dumps({"doc": schedule_data}), refresh=True)

        except Exception as e:
            print(e)
            return {"failure": str(e)}

File: frontend/static/js/scheduler.js
/*
   Copyright 2022 Jerome DE LUCCHI

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
import FrontendConfig from './frontend.js'
var config = new FrontendConfig()

let Scheduler = class {

    constructor() {

        this._id = undefined
        this._name = undefined
        this._description = undefined
        this._scenarios = undefined
        this._daily = undefined
        this._status = undefined
        this._time = undefined
        this._interval = undefined
        this._realm = undefined
        this._data = undefined
    }

    set id(id) { this._id = id }
    set name(name) { this._name = name }
    set description(description) { this._description = description }
    set scenarios(scenarios) { this._scenarios = scenarios }
    set daily(daily) { this._daily = daily }
    set status(status) { this._status = status }
    set time(time) { this._time = time }
    set interval(interval) { this._interval = interval }
    set realm(realm) { this._realm = realm }
    set rawData(data) { this._data = data}

    get id() { return this._id }
    get name() { return this._name }
    get description() { return this._description }
    get scenarios() { return this._scenarios }
    get daily() { return this._daily }
    get time() { return this._time }
    get interval() { return this._interval }
    get realm() { return this._realm }
    get rawData() { return this._data}

    add = async (formData) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/schedulers'
        let header = { 'Authorization': config.session.httpToken }
        let data = JSON.stringify(formData)
        let response = await fetch(url, {method: "POST", headers: header, body: data})
        if (response.ok) {
            response = JSON.parse(await response.text())
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    list = async () => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/schedulers'
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: "GET", headers: header})
        if (response.ok) {
            response = JSON.parse(await response.text())
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    listById = async (schedulerId) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/schedulers/' + schedulerId
        let header = { 'Authorization': config.session.httpToken }
        let response = await fetch(url, {method: "GET", headers: header})
        if (response.ok) {
            response = JSON.parse(await response.text())
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    action = async (scheduler_data) => {
        let url = config.proxyAPI + '/realms/' + config.session.realm + '/schedulers/action'
        let header = { 'Authorization': config.session.httpToken, 'Content-Type': "application/json; charset=utf-8"}
        let data = JSON.stringify(scheduler_data)
        let response = await fetch(url, {method: "POST", headers: header, body: data})
        if (response.ok) {
            response = JSON.parse(await response.text())
            if (Object.keys(response).includes("tokenExpired")) { account.logout() } else { return response }
        }
    }

    load = (schedulerData) => {
        this.id = schedulerData["_id"]
        this.name = schedulerData["_source"]["name"]
        this.description = schedulerData["_source"]["description"]
        this.scenarios = schedulerData["_source"]["scenarios"]
        this.daily = schedulerData["_source"]["daily"]
        this.status = schedulerData["_source"]["status"]
        this.time = schedulerData["_source"]["time"]
        this.interval = schedulerData["_source"]["interval"]
        this.realm = schedulerData["_source"]["realm"]
        this.rawData = schedulerData["_source"]
    }
}

export default Scheduler